src/expr.rs

```diff
@@ -115,11 +116,99 @@ impl Rewrite for ast::Expr {
                 };
                 Some(format!("continue{}", id_str))
             }
+            ast::Expr_::ExprClosure(capture, ref fn_decl, ref body) => {
+                rewrite_closure(capture, fn_decl, body, self.span, context, width, offset)
+            }
             _ => context.codemap.span_to_snippet(self.span).ok(),
         }
     }
 }
 
+// This functions is pretty messy because of the wrapping and unwrapping of
+// expressions into and from blocks. See rust issue #27872.
+fn rewrite_closure(capture: ast::CaptureClause,
+                   fn_decl: &ast::FnDecl,
+                   body: &ast::Block,
+                   span: Span,
+                   context: &RewriteContext,
+                   width: usize,
+                   offset: usize)
+                   -> Option<String> {
+    let mover = if capture == ast::CaptureClause::CaptureByValue {
+        "move "
+    } else {
+        ""
+    };
+    let offset = offset + mover.len();
+
+    // 4 = "|| {".len(), which is overconservative when the closure consists of
+    // a single expression.
+    let argument_budget = try_opt!(width.checked_sub(4 + mover.len()));
+    // 1 = |
+    let argument_offset = offset + 1;
+
+    let arg_items = itemize_list(context.codemap,
+                                 fn_decl.inputs.iter(),
+                                 "|",
+                                 |arg| span_lo_for_arg(arg),
+                                 |arg| span_hi_for_arg(arg),
+                                 |arg| rewrite_fn_input(arg),
+                                 span_after(span, "|", context.codemap),
+                                 body.span.lo);
+
+    let fmt = ListFormatting::for_fn(argument_budget, argument_offset);
+    let prefix = format!("{}|{}|", mover, write_list(&arg_items.collect::<Vec<_>>(), &fmt));
+    let block_indent = closure_block_indent(context, offset);
+
+    let body_rewrite = if body.stmts.is_empty() {
+        let expr = body.expr.as_ref().unwrap();
+        // All closure bodies are blocks in the eyes of the AST, but we may not
+        // want to format them as such when they only contain a single
+        // expression.
+        let inner_expr = match expr.node {
+            ast::Expr_::ExprBlock(ref inner_body) if inner_body.stmts.is_empty() => {
+                inner_body.expr.as_ref().unwrap()
+            }
+            _ => expr,
+        };
+
+        // 1 = the separating space between arguments and the body.
+        let extra_offset = extra_offset(&prefix, offset) + 1;
+        let rewrite = inner_expr.rewrite(context, width - extra_offset, offset + extra_offset);
+
+        // Checks if rewrite succeeded and fits on a single line.
+        let accept_rewrite = rewrite.as_ref().map(|result| !result.contains('\n')).unwrap_or(false);
+
+        if accept_rewrite {
+            rewrite
+        } else {
+            if let ast::Expr_::ExprBlock(ref inner_body) = expr.node {
```

Comment:
I don't understand the logic here. Why try and rewrite inner_expr if it is a multi-stmt block? (btw what happens if it is a double nested block - do we break our idempotence invariant then?). It seems that you should handle two cases (single expr vs multi-stmt block) at a higher level and then if the single expr case fails, wrap the AST node in a block and then run the block path, rather than doing it manually. Would that work? (I.e., all manipulation is done at the AST level rather than the text level).